akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -60,11 +59,21 @@ final class Merge[T] private (val inputPorts: Int, val eagerComplete: Boolean) e
 
     override def preStart(): Unit = in.foreach(tryPull)
 
+    @tailrec
     private def dequeueAndDispatch(): Unit = {
       val in = pendingQueue.dequeue()
-      push(out, grab(in))
-      if (upstreamsClosed && !pending) completeStage()
-      else tryPull(in)
+      if (in == null) {
+        // in is null if we reached the end of the queue
+        if (upstreamsClosed) completeStage()
+      } else if (isAvailable(in)) {
```

Comment:
I changed it as it being closed is the reason for discarding it when going through the queue.